File: codes_algo/code_C/quickSort.c

/* ----------------------

Name: 	Quicksort algorithm.
Author: Ranaji Krishna.

Notes:	The fxn. sorts arrays in an ascending order.

Inputs: Pointer to the array to be sorted, starting index (=0)
	and size of array

Outputs: Pointer to the sorted array.
------------------------- */

#include <stdio.h>
#include <stdlib.h>

int partition(int *tmpArray, int st, int ed){
	
	int pvt = *(tmpArray + ed);	// Set the pivot.	
	int int1 = st;			// Starting index.

	while(st < ed){
		if(*(tmpArray+st) < pvt){		// Swap if value is less than the pivot.
			int m = *(tmpArray+int1);	
			*(tmpArray + int1) = *(tmpArray + st);
			*(tmpArray + st) = m;
			int1++;
		}
	st++;
	}

	*(tmpArray + ed) = *(tmpArray + int1);		// Update starting index.
	*(tmpArray + int1) = pvt;			// Update pivot.

	return(int1);
}


int *quick_sort(int *pArray, int strt, int end){

	if(strt < end){					// Sort.
		int int1;
		int1 = partition(pArray, strt, end);
		pArray = quick_sort(pArray, strt, int1 - 1);
		pArray = quick_sort(pArray, int1 + 1, end);
	}
	return(pArray);
}

File: codes_algo/code_R/quick_sort.R


partition <- function(tmpArray, st, ed){
  pvt <- tmpArray[ed];
  int1 <- st;
  while(st < ed){
    if(tmpArray[st] < pvt){
      m <- tmpArray[int1];
      tmpArray[int1] <- tmpArray[st];
      tmpArray[st] <- m;
      int1 <- int1 + 1;
    }
    st <- st + 1;
  }
  tmpArray[ed] <- tmpArray[int1];
  tmpArray[int1] <- pvt;
  return(list(tmpArray, int1));
}

quick_sort <- function(tmpArray, st, ed){
  if(st < ed){ 
    strArray <- partition(tmpArray, st, ed);
    tmpArray <- strArray[[1]]; int1 <- strArray[[2]];
    tmpArray <- quick_sort(tmpArray, st, int1 - 1);
    tmpArray <- quick_sort(tmpArray, int1 + 1, ed);
  }
  return(tmpArray)  
}

# set.seed(10);
# a <- as.array(round(runif(10,0,10)));
# b <- quick_sort(a,1,dim(a));
# print(a)
# print(b)
